### PR TITLE
WIP: Use assign expr for "while True: ... break"

### PR DESCRIPTION
**Pull request by vstinner**
Replace:

```
    while True:
        x = expr
        if not x:
            break
        ...
```

with:

```

    while (x := expr):
        ...
```

Notes:

* Corner case: (x := expr) and (x != NUL), use x in the second test,
  whereas x is set in the first test
* asyncio/sslproto.py change avoids adding an empty chunk to appdata.
* some changes remove comments
* some lines are longer than 80 characters to keep everything on the
  same line

### CODE REVIEW

**Comment by kshal88** [comment]
Delete

File: Lib/_pyio.py [hunk 1]
```diff
@@ -588,10 +588,7 @@ def read(self, size=-1):
     def readall(self):
         """Read until EOF, using multiple read() call."""
         res = bytearray()
-        while True:
-            data = self.read(DEFAULT_BUFFER_SIZE)
-            if not data:
-                break
+        while (data := self.read(DEFAULT_BUFFER_SIZE)):
```
Comment (JelleZijlstra) [inline Lib/_pyio.py:591]: Why are you using parentheses here? I don't think they are syntactically necessary.

Comment (vstinner) [reply Lib/_pyio.py:591]: I'm too lazy to read the latest PEP, and the current implementation of the PEP is outdated, so I added parenthesis even if they are not needed.

**Comment by vstinner** [comment]
> Delete

The point of this PR is just to open discuss on coding style: discuss when assignment expressions are appropriate or not.

The discussion: "[Python-Dev] Assignment expression and coding style: the while True case"
https://mail.python.org/pipermail/python-dev/2018-July/154323.html

"Let's say that the PEP 572 (assignment expression) is going to be approved. Let's move on and see how it can be used in the Python stdlib." :-)

I added the DO-NOT-MERGE label and "WIP" in the title. I don't want to merge this PR (at least, not as it is currently).

**Comment by vstinner** [comment]
> I hope the parenthesis aren't necessary though as they do feel a little unpythonic

Parenthesis are optional. I chose to always add them.

I don't think that we can already say if parenthesis are pythonic or not since it's a new syntax :-) We don't have enough experience with it :-)

**Comment by liiight** [comment]
I meant that using parenthesis in if statements is usually redundant and I'd be happy to match that behaviour with the new syntax

**Comment by tim-one** [comment]
The PEP explicitly rejected requiring parentheses, giving as one concrete example that they'd "look redundant" in:

    # Top level in if
    if match := pattern.match(line):
        return match.group(1)

Which is exactly the kind of case in which you're always adding them and are getting feedback that it just doesn't look right to experienced Python eyes.  I didn't write that part or the PEP, or even participate in the discussion that resulted in it being added - I just find it to be obviously true 😉 

I can't think of any case in which anyone recommends adding redundant parentheses _except_ when clarifying evaluation order.

For example, I doubt anyone would object to the redundant parens in:

    i = 1 << (2 + 1)
or

    t = (1, 2)

But I can't think either of any experienced Python programmer who does champion the redundant parens in

    i = (1 + 2)
    i = (((1) + (2)))
    return (1 + 2)
    if (x or y):
    while (x > y):
    ...

etc.  The parens don't _clarify_ anything in those - they just add counterproductive line noise.  Newcomers from other languages where parens are required in such contexts often start off adding them anyway, and that's fine, but most grow out of it soon enough even if they're not slapped for it in a code review 😉 

You started off by saying you wanted a discussion of coding guidelines for use of `:=`.  Well, that's one you're getting consistent feedback on:  the same as with any other bit of Python expression syntax, don't add redundant line noise unless it serves a _purpose_ (which, for parens, is pretty much limited to forcing or clarifying evaluation order in expressions with multiple operators).

"They're optional" is no reason _for_ adding them; "they're optional" precisely in recognition of that adding them serves no _purpose_ in these common cases.  They're just noise here.

File: Lib/re.py [hunk 2]
```diff
@@ -349,13 +349,7 @@ def scan(self, string):
...
         match = self.scanner.scanner(string).match
         i = 0
-        while True:
-            m = match()
-            if not m:
-                break
-            j = m.end()
-            if i == j:
-                break
+        while (m := match()) and (j := m.end()) == i:
```
Comment (JimJJewett) [inline Lib/re.py:352]: Doesn't that reverse the test on j?  As in, shouldn't it be 
while (m := match()) and (j := m.end()) != i:

Though I'm not sure which way this argues; the original logic was at least as hard to follow as the new code.

File: Lib/binhex.py [hunk 3]
```diff
@@ -237,16 +237,12 @@ def binhex(inp, out):
 
     with io.open(inp, 'rb') as ifp:
         # XXXX Do textfile translation on non-mac systems
-        while True:
-            d = ifp.read(128000)
-            if not d: break
+        while (d := ifp.read(128000)):
```
Comment (JimJJewett) [inline Lib/binhex.py:240]: This is a bigger win than it looks like.

Note that the original "if not d: break" on a single line was itself arguably a style violation because the break doesn't start a line, even though it might change flow-control.  That said, I would have done the same thing as the original author to keep it short; it feels like a guard instead of a real control "path".

So this is an area where vertical space matters, the "while True: " is ugly, and the current spelling is already doing slightly ugly things to avoid drawing extra emphasis ... and shortening it with the inline assignment improves on all three.

File: Lib/sre_parse.py [hunk 4]
```diff
@@ -507,24 +507,17 @@ def _parse(source, state, verbose, nested, first=False):
...
             # skip whitespace and comments
             if this in WHITESPACE:
                 continue
             if this == "#":
-                while True:
-                    this = sourceget()
-                    if this is None or this == "\n":
-                        break
+                while (this := sourceget()) is not None and (this != "\n"):
+                    pass
```
Comment (JimJJewett) [inline Lib/sre_parse.py:520]: In this case, you don't even need "this" again, because the next instruction continues the enclosing loop, which resets this.  (Essentially, "this" was a dummy name, getting re-used.)

    while sourceget() not in (None, "\n"): pass

File: Lib/tarfile.py [hunk 5]
```diff
@@ -483,15 +483,11 @@ def _init_read_gz(self):
             xlen = ord(self.__read(1)) + 256 * ord(self.__read(1))
             self.read(xlen)
         if flag & 8:
-            while True:
-                s = self.__read(1)
-                if not s or s == NUL:
-                    break
+            while (s := self.__read(1)) and s != NUL:
```
Comment (JimJJewett) [inline Lib/tarfile.py:486]: I know the second part (s != NUL) is controversial -- but consider this one vote saying "yes, this was still a good replacement".  The meaning behind the line is "throw away characters until you get to the end of the string", and splitting the test into two parts is just an distraction (needed because of how the end of string happens to be defined for this function).  

The existing 4 lines make it seem bigger than it is, but so would the three-line alternative suggested by Tim.  Frankly, even the 2nd line for "pass" seems slightly misleading, but I don't see a further improvement.  

Also note that s is really a dummy variable, only needed to allow comparisons against a 2nd type of string-end.  Same comments for flag&16 -- if there is a difference in the cases, the length caused me to miss it.

File: Lib/tarfile.py [hunk 6]
```diff
@@ -520,10 +516,7 @@ def read(self, size=None):
         """
         if size is None:
             t = []
-            while True:
-                buf = self._read(self.bufsize)
-                if not buf:
-                    break
+            while (buf := self._read(self.bufsize)):
```
Comment (JimJJewett) [inline Lib/tarfile.py:519]: https://github.com/python/cpython/commit/8d130913cb9359c01de412178f9942419e921170  has eliminated this code (and just asserted that users will always call with a non-None size.)

File: Lib/tarfile.py [hunk 7]
```diff
@@ -2339,10 +2325,8 @@ def _load(self):
         """Read through the entire archive file and look for readable
            members.
         """
-        while True:
-            tarinfo = self.next()
-            if tarinfo is None:
-                break
+        while (tarinfo := self.next()) is not None:
+            pass
```
Comment (JimJJewett) [inline Lib/tarfile.py:2329]: This is another controversial change that I like.  The first one looks like it is doing some sort of work to select or filter; the 2nd makes it more clear that it is just skipping through (but touching and therefore loading?) all the self.next() chunks.  

That said, the tarfile temp name is not used again.  It might be even more clear as

while self.next() is not None: pass

**Review by JimJJewett — COMMENTED** [review]
I consider these all wins, particularly where the original author had written if (whatever): break on a single line.

**Comment by vstinner** [comment]
The PEP 572 has been accepted. I only wrote these WIP PEPs to help *me* to discuss the PEP. I never intended to propose these changes to be merged. So I close this PR.

File: Lib/telnetlib.py [hunk 8]
```diff
@@ -562,10 +562,7 @@ def mt_interact(self):
         """Multithreaded version of interact()."""
         import _thread
         _thread.start_new_thread(self.listener, ())
-        while 1:
-            line = sys.stdin.readline()
-            if not line:
-                break
+        while (line := sys.stdin.readline()):
             self.write(line.encode('ascii'))
```
Comment (mgedmin) [inline Lib/telnetlib.py:566]: TBH `for line in sys.stdin:` seems better to me here.
